### PR TITLE
Save useful links shared in slack

### PR DESCRIPTION
**Pull request by ujjwal96**
This PR adds a save links handler which can be used to save links shared in slack.
To setup - You can fork or copy the files from https://github.com/ujjwal96/links (also follow the readme steps there) to OTA org on github and edit the variables `GITHUB_REPO`, `GITHUB_BRANCH` in `save_handler.py`  
To save links users can comment `!link <category>` on any message containing a URL.

Demo - https://ujjwal96.github.io/links/

### CODE REVIEW

File: handlers/save_handler.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from util.savelinkhelper import unfurl
+
+GITHUB_REPO = ""  # owner/repo-name
+GITHUB_BRANCH = ""  # e.g master
+
+CATEGORIES = ["web", "pwn", "re", "crypto", "misc"]
+
+
+class SaveCommand(Command):
+    """Save a url from a slack message according to a specific category"""
```
Comment (Grazfather) [inline handlers/save_handler.py]: Period at the end

File: handlers/save_handler.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    @classmethod
+    def execute(cls, slack_wrapper, args, timestamp, channel_id, user_id, user_is_admin):
+        """Execute the save link command"""
```
Comment (Grazfather) [inline handlers/save_handler.py]: save command, not save link. Add period.

File: handlers/save_handler.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        if not GITHUB_REPO and not GITHUB_BRANCH:
```
Comment (Grazfather) [inline handlers/save_handler.py]: use OR not AND, since both have to be set.

File: handlers/save_handler.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            raise InvalidCommand("Link saver not configured")
+        if args[0] not in CATEGORIES:
+            raise InvalidCommand("Invalid Category")
+
+        message = slack_wrapper.get_message(channel_id, timestamp)["messages"][0]["text"]
+        url_regex = "((https?):\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?"
+        url = re.search(url_regex, message)
+
+        if not url:
+            slack_wrapper.post_message(channel_id, "Couldn't extract URL", timestamp)
```
Comment (Grazfather) [inline handlers/save_handler.py]: Look at how errors are posted as replies (in a thread) this is cleaner.

File: handlers/save_handler.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            return
+
+        url_data = unfurl(url.group())
+
+        data = {
+            "fields[title]": url_data["title"],
+            "fields[link]": url.group(),
+            "fields[description]": url_data["desc"],
+            "fields[category]": "{}".format(args[0])
```
Comment (Grazfather) [inline handlers/save_handler.py]: `string(args[0])` or even just `args[0]`. Why use `.format`?

File: handlers/save_handler.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        resp = requests.post(
+            "https://dev.staticman.net/v3/entry/github/{}/{}/links".format(
+                GITHUB_REPO,
+                GITHUB_BRANCH
+            ),
+            data=data
+        ).json()
+
+        if resp["success"]:
```
Comment (Grazfather) [inline handlers/save_handler.py]: does this key always exist in the response?

File: handlers/save_handler.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            slack_wrapper.post_message(channel_id, "Link saved successfully", timestamp)
+        else:
+            slack_wrapper.post_message(channel_id, "Error saving the link: `{}`".format(resp), timestamp)
+
+
+class SaveHandler(BaseHandler):
+    """Handler for saving links"""
```
Comment (Grazfather) [inline handlers/save_handler.py]: period

File: util/savelinkhelper.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        return desc["content"].strip()
+    else:
+        return ""
+
+
+def unfurl(url: str):
+    details = {}
+    resp = requests.get(url).text
+    soup = BeautifulSoup(resp, "html.parser")
+    details["title"] = get_title(soup)
```
Comment (Grazfather) [inline util/savelinkhelper.py]: prefer

```python
details = {
    "title":...,
    "desc":...,
}
```

File: handlers/save_handler.py (hunk 5 above)
Comment (ujjwal96) [reply handlers/save_handler.py]: I should probably sleep a little :sweat_smile:

File: handlers/save_handler.py (hunk 6 above)
Comment (ujjwal96) [reply handlers/save_handler.py]: I believe it does

File: handlers/save_handler.py (hunk 4 above)
Comment (ujjwal96) [reply handlers/save_handler.py]: So, something like "Error: `unable to extract URL`" ?

**Comment by Kileak** [comment]
Cannot look through this completely today, but some short questions:

- What's the advantage of using staticman instead of using the existing githandler for uploading the jekyll posts (like it's done in SolveTracker)
- I see some potential for abuse in this command, might make sense to add an "allowed user" list or something for administrating those links after commit (or maybe before they're uploaded)?

**Comment by ujjwal96** [comment]
> What's the advantage of using staticman instead of using the existing githandler for uploading the jekyll posts (like it's done in SolveTracker)

Well, you don't need to setup the other repo locally (might save some storage on the server). If you feel strongly against it then I can change it to use the githandler

> I see some potential for abuse in this command, might make sense to add an "allowed user" list or something for administrating those links after commit (or maybe before they're uploaded)?

Staticman doesn't directly push to the repo, it sends a PR which can then be merged. I'll try to see if we can have some kind of authentication with this

**Comment by Kileak** [comment]
No, I'm not strongly against using it, just interested in the advantages. 

But common functionality like this, should be moved into an own class, so it can be reused (for example, if this approach makes more sense than the existing handler, we could also use it for SolveTracker). I can also do this later on, though.

I'll try to test this in the test slack this week to see, how it works out. 

Though, I feel, we might get more out of this, if we store the stuff in an extra database and make that one also searchable via slack (via keywords) or something similar. 

Maybe we can do both, so everyone can search for those links in the way he prefers more (I just think, the "link" page might get pretty bloated after some time, if everyone is just passing links to it)

**Comment by ujjwal96** [comment]
I'll move this into its own class :+1: 
For the external db, let's have a conversation in #botdev and figure out how to approach it :slightly_smiling_face:

**Comment by ujjwal96** [comment]
> we should just add a list of users that are allowed to push via the bot (like the admin_handler does it)
adding people to this list should just be an admin-command (that can be done via a flag on the command descriptor)
so admins can decide who is in the "link-poster" group, (the list can just be stored in the config.json)
and only those people are allowed to use the savelink command

From Kileak on slack

**Pull request merged**